test/e2e/xfn_test.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2023 The Crossplane Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"math/big"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/vladimirvivien/gexe"
+	"golang.org/x/exp/slices"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	client2 "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+	"sigs.k8s.io/e2e-framework/third_party/helm"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+
+	v1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+	"github.com/crossplane/crossplane/test/e2e/funcs"
+)
+
+func TestFunctions(t *testing.T) {
+
+	manifests := "test/e2e/manifests/apiextensions/composition/functions/private-registry"
+	environment.Test(t,
+		features.New("PullImageFromPrivateRegistryWithCustomCert").
+			WithLabel(LabelArea, "xfn").
+			Setup(funcs.CreateNamespace("reg")).
+			Setup(CreateTLSCertificateAsSecret("private-docker-registry.reg.svc.cluster.local", "reg")).
+			Setup(InstallDockerRegistry()).
+			Setup(CopyImagesToRegistry()).
+			Setup(CrossplaneDeployedWithFunctionsEnabled()).
+			Setup(ProvideNopDeployed()).
+			Assess("CompositionWithFunctionIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "composition.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "composition.yaml"),
+			)).
+			Assess("ClaimIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "claim.yaml"),
+			)).
+			Assess("ClaimBecomesAvailable", funcs.ResourcesHaveConditionWithin(5*time.Minute, manifests, "claim.yaml", xpv1.Available())).
+			Assess("ManagedResourcesProcessedByFunction", ManagedResourcedProcessedByFunction()).
+			// Teardown(funcs.AsFeaturesFunc(funcs.HelmUpgrade(HelmOptions()...))).
```

Comment:
A few things to make this similar to the design [proposal](https://github.com/crossplane/crossplane/pull/4270/files#diff-7d9a55ec9712c1ccc931efe126dbea316dab465e6ab265abd089736be5511f47R88) we're going to merge soon:
- Use `WithSetup` instead of `Setup` (same for `Teardown`), as that allows to specify names for setup steps too.
- Avoid defining helper functions here, we'll refactor tests by adding helpers to the `funcs` package once we start having more tests and seeing patterns. So please, if required add an helper there, otherwise, for simple wrappers around a single `funcs.AllOf` such as `CrossplaneDeployedWithFunctionsEnabled` and `ProvideNopDeployed`, etc..., or around single functions such as `InstallDockerRegistry`.
- Avoid relying on another test's manifests, as you are doing in ProvideNopDeployed (which being hidden in an helper function, adding to the point above, was not so easy to spot), as that would make the tests brittle